ports/stm32/omv_csi.c

```diff
@@ -611,13 +617,43 @@ static int stm_csi_snapshot(omv_csi_t *csi, image_t *image, uint32_t flags) {
                 // Special transfer mode that uses DMA in circular mode and MDMA
                 // to move the lines to the final destination.
                 ((DMA_Stream_TypeDef *) csi->dma.Instance)->CR |= DMA_SxCR_CIRC;
+                csi->one_shot = true;
                 HAL_DCMI_Start_DMA(&csi->dcmi, DCMI_MODE_CONTINUOUS, (uint32_t) &_line_buf, line_width / 4);
             #endif  // USE_MDMA
             } else {
                 #if defined(STM32F4) || defined(STM32F7) || defined(STM32H7)
                 // Start a multibuffer (line by line) transfer.
                 HAL_DCMI_Start_DMA_MB(&csi->dcmi, DCMI_MODE_CONTINUOUS, (uint32_t) &_line_buf, csi->dma_size, fb->v);
                 #else
+                // Handle YUV422 Source -> Y Destination using DCMI byte drop.
+                if ((csi->pixformat == PIXFORMAT_GRAYSCALE) && (csi->mono_bpp == 2)) {
+                    csi->dma_size /= 2;
+                    csi->dcmi.Instance->CR |= DCMI_CR_BSM_0;
+                } else {
+                    csi->dcmi.Instance->CR &= ~DCMI_CR_BSM_0;
+                }
+
+                // Turn on/off byte swapping for RGB/YUV formats.
+                for (size_t i = 0; i < OMV_ARRAY_SIZE(dma_nodes); i++) {
+                    if ((csi->pixformat == PIXFORMAT_RGB565 && csi->rgb_swap) ||
+                        (csi->pixformat == PIXFORMAT_YUV422 && csi->yuv_swap)) {
+                        dma_nodes[i].LinkRegisters[NODE_CTR1_DEFAULT_OFFSET] |= DMA_CTR1_DBX;
+                    } else {
+                        dma_nodes[i].LinkRegisters[NODE_CTR1_DEFAULT_OFFSET] &= ~DMA_CTR1_DBX;
+                    }
+                }
+
+                // HAL_DCMI_Start_DMA() for transfer sizes less than 64KB expects a non-circular linked list.
+                if ((csi->dma_size * 4) <= (OMV_CSI_DMA_MAX_SIZE / 4)) {
+                    // Modify the link list head to point to nothing.
+                    dma_nodes[0].LinkRegisters[NODE_CLLR_LINEAR_DEFAULT_OFFSET] = 0;
+                } else {
+                    // Modify the link list head to point to the tail node (default).
+                    dma_nodes[0].LinkRegisters[NODE_CLLR_LINEAR_DEFAULT_OFFSET] =
+                        DMA_CLLR_UT1 | DMA_CLLR_UT2 | DMA_CLLR_UB1 | DMA_CLLR_USA | DMA_CLLR_UDA | DMA_CLLR_ULL |
+                        (((uint32_t) &dma_nodes[1]) & DMA_CLLR_LA);
+                }
```

Comment:
```suggestion
                // Disable circular mode for transfer sizes less than 64KB.
                if (csi->dma_size <= OMV_CSI_DMA_MAX_SIZE / 4) {
                    HAL_DMAEx_List_ClearCircularMode(&dma_queue);
                } else {
                    HAL_DMAEx_List_SetCircularMode(&dma_queue);
                }
```
The test seems to off to me. `dma_size` is number of elements, `OMV_CSI_DMA_MAX_SIZE` is bytes so either multiply `dma_size` or divide `OMV_CSI_DMA_MAX_SIZE` but not both.

About all the extra parentheses put everywhere:

https://en.cppreference.com/w/c/language/operator_precedence.html